pallets/proxy-bonding/src/functions.rs

```diff
@@ -0,0 +1,112 @@
+use frame_support::ensure;
+use crate::{AccountIdOf, AssetId, BalanceOf, Config, Pallet, ReleaseType, Releases};
+use frame_support::traits::{
+	fungible,
+	fungible::{Inspect, Mutate, MutateHold},
+	fungibles,
+	fungibles::Mutate as FungiblesMutate,
+	tokens::{Fortitude, Precision, Preservation},
+};
+use frame_system::pallet_prelude::BlockNumberFor;
+use polimec_common::ProvideAssetPrice;
+use sp_runtime::{
+	traits::{AccountIdConversion, Get},
+	FixedPointNumber,
+};
+
+impl<T: Config> Pallet<T> {
+	/// Calculate the USD fee in `fee_asset` for bonding `bond_amount` of the native token.
+	/// e.g. if the fee is 1%, native token PLMC, fee_asset USDT, bond_amount 1000 PLMC, PLMC price 0.5USD, USDT price 1USD,
+	/// Then the calculated fee would be 1% * 1000 * 0.5 = 5USD, which is 5 USDT at a price of 1USD.
+	pub fn calculate_fee(bond_amount: BalanceOf<T>, fee_asset: AssetId) -> Result<BalanceOf<T>, &'static str> {
+		let bonding_token_price = T::PriceProvider::get_decimals_aware_price(
+			T::BondingTokenId::get(),
+			T::UsdDecimals::get(),
+			T::BondingTokenDecimals::get(),
+		)
+		.ok_or("Price not available")?;
```

Comment:
Is there a specific reason why we use `&'static str` as an error and not a `DispatchError`?